cpp/src/generators/generator_tools.cu

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cugraph/graph_generators.hpp>
+#include <cugraph/utilities/error.hpp>
+#include <experimental/scramble.cuh>
+
+#include <raft/cuda_utils.cuh>
+
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/execution_policy.h>
+#include <thrust/for_each.h>
+
+#include <numeric>
+
+namespace cugraph {
+
+namespace detail {
+
+template <typename T>
+void append_all(raft::handle_t const &handle,
+                std::vector<rmm::device_uvector<T>> &&input,
+                rmm::device_uvector<T> &output)
+{
+  auto output_iter = output.begin();
+
+  for (size_t i = 0; i < input.size(); ++i) {
+    raft::copy(output_iter, input[i].begin(), input[i].size(), handle.get_stream());
+    output_iter += input[i].size();
+  }
+}
+
+}  // namespace detail
+
+template <typename vertex_t>
+void scramble_vertex_ids(raft::handle_t const &handle,
+                         rmm::device_uvector<vertex_t> &d_src_v,
+                         rmm::device_uvector<vertex_t> &d_dst_v,
+                         vertex_t vertex_id_offset,
+                         uint64_t seed)
```

Comment:
Have you checked that this function works if num_vertices is not a power of two?

IIRC, this function works if vertices are in the range [0, 2^scale), but I am not sure this still works otherwise. Need to check.

Reply:
I guess this will still work (but the scrambled vertex IDs are no longer contiguous integers but does not matter if we renumber or allow having many isolated vertices... but needs to double check).

Reply:
Added a unit test for the scramble function.  The result is correct (it validates properly).  Your assertion is correct regarding the resulting ids no longer being contiguous.  This will (as you observe) result in a collection of isolated vertices if we do not renumber; although if we renumber that will be corrected.